src/subcommands/publishers.rs

```diff
@@ -3,48 +3,99 @@ use std::collections::{BTreeMap, HashMap};
 use crate::publishers::fetch_owners_of_crates;
 use crate::{common::*, publishers::PublisherData};
 
-pub fn publishers(args: Vec<String>, max_age: std::time::Duration) -> Result<(), std::io::Error> {
+use serde::{Deserialize, Serialize};
+
+use std::collections::hash_map::Entry;
+
+#[derive(Serialize, Deserialize)]
+struct PublishersJdoc {
+    warning: String,
+    complaint: String,
+    publishers: HashMap<String, Vec<String>>,
+}
+
+pub fn publishers(
+    args: Vec<String>,
+    max_age: std::time::Duration,
+    output_json: bool,
+) -> Result<(), std::io::Error> {
     let dependencies = sourced_dependencies(args);
-    complain_about_non_crates_io_crates(&dependencies);
+    let complaint = complain_about_non_crates_io_crates(&dependencies);
     let (publisher_users, publisher_teams) = fetch_owners_of_crates(&dependencies, max_age)?;
 
-    if !publisher_users.is_empty() {
-        println!("\nThe following individuals can publish updates for your dependencies:\n");
-        let user_to_crate_map = transpose_publishers_map(&publisher_users);
-        let map_for_display = sort_transposed_map_for_display(user_to_crate_map);
-        for (i, (user, crates)) in map_for_display.iter().enumerate() {
-            // We do not print usernames, since you can embed terminal control sequences in them
-            // and erase yourself from the output that way.
-            // TODO: check if it's possible to smuggle those into github/crates.io usernames
-            let crate_list = comma_separated_list(&crates);
-            println!(" {}. {} via crates: {}", i + 1, &user.login, crate_list);
+    if output_json {
+        let mut publishers_jdoc = PublishersJdoc {
+            warning: concat!(
+                "Note: there may be outstanding publisher invitations.",
+                "crates.io provides no way to list them. ",
+                "See https://github.com/rust-lang/crates.io/issues/2868 for more info."
+            )
+            .to_string(),
+            complaint: complaint,
+            publishers: HashMap::new(),
+        };
+
+        for list in vec![publisher_users, publisher_teams].iter() {
+            for (crate_name, publishers) in list.iter() {
```

Comment:
It's not necessary to allocate here. Rather, we can chain them or iterate over an array.

```suggestion
        for (crate_name, publishers) in publisher_users.iter().chain(&publisher_teams) {
```